Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+event's keyword and the event's level have been enabled on the provider to which the event belongs.
+
+To accomplish this, the GC will contain a class with this signature:
+
+```c++
+// Constructed on GC startup for every event that the GC can fire.
+struct GCEvent
+{
+    /* opaque, but will contain information about what event this is,
+       its keyword, its level, and a handle to the EE's representation of this event */
+    /* this structure is potentially auto-generated from an event manifest */
+};
+
+class GCEventStatus
+{
+public:
+    // Returns true if this event is currently enabled, false otherwise.
+    static bool IsEventEnabled(const GCEvent& event);
+
+    // Enables events with the given keyword and level.
+    static void Enable(int keyword, int level);
+
+    // Disables events the given keyword and level.
+    static void Disable(int keyword, int level);
+};
+
+```
+
+The GC will use `GCEventStatus::IsEventEnabled` to query whether or not a particular event is enabled.
+Whenever the EE observes a change in what keywords or levels are enabled, it must inform the GC of the
+change so that it can update `GCEventStatus` using `Enable` and `Disable`. The exact mechanism by which
+the EE observes a change in the event state is described further below. ("Getting Informed of Changes to 
+Event State")
+
+When the EE *does* observe a change, it informs the GC using these two new methods on `IGCHeap`:
+
+```c++
+void IGCHeap::EnableEvents(int keyword, int level)
+void IGCHeap::DisableEvents(int keyword, int level)
+```
+
+
+These interface methods in turn will call `GCEventStatus::{Enable, Disable}` in order
+to enable or disable keywords or levels. The currently enabled keywords and levels are encoded as
+bit vectors so that querying whether an event is enabled is as fast as loading three globals, a pointer
+dereference, and two bit tests, e.g. (hypothetically):
+
+```c++
+uint32_t enabledLevels;
+uint32_t enabledKeywords;
+
+bool GCEventStatus::IsEventEnabled(const GCEvent& event)
+{
+    return (enabledLevels & event.level) && (enabledKeywords & event.keyword);
+}
+```
+
+## Firing Events
+
+In order to fire an event, the GC will need to communicate with the EE in some way. The EE is ultimately
+responsible for routing the event to any appropriate subsystems (ETW, LTTNG, EventPipe) and the GC has
+no knowledge of what it is going to do with events that we give it.
+
+Given that we have validated that an event is enabled (through `GCEventStatus::IsEventEnabled`), we can
+fire the event using this new method on `IGCToCLR`:
+
+```c++
+void IGCToCLR::FireKnownEvent(
+    /* IN */ void* eventHandle,
+    /* IN */ void** eventData,
+    /* IN */ uint32_t eventDataCount,
+    /* IN */ const GUID* pActivityId = nullptr,
+    /* IN */ const GUID* pRelatedActivityId = nullptr
+);
+```
+
+The four arguments of this function are:
+1. `eventHandle` - An opaque handle to the EE's representation of the event being fired.
```

Comment:
How does the GC gets this handle?

Reply:
They must be provided to the GC on startup. This can either be done as a callback (in the same vein as `IGCToCLR::GetFreeObjectMethodTable`) or done on startup as an argument to `GC_Initialize`. I think I'm leaning towards the former, if we go this route; it seemed like a bad idea to me to add more parameters to `GC_Initialize` than the bare minimum of what is needed to initialize GC functionality.

Reply:
This fells way too complicated - event handles that the GC keeps track of, dummy arguments filled in by the EE, ... .

Would it be better to have an event tracing interface that has one method per event? It would make the tracing contract simple and explicit.

Reply:
The GC fires a lot of events so that would be a lot of callbacks. If that's a cleaner solution, i'm all for it! I am also wary of adding lots of new methods to the interface, though.

Reply:
It would be an option to have a tracing interface and just add a new function to it every time a new event is added.  If this is done, the work to fire the event is essentially just boiler-plate - on the EE side just call FireEtw* and the EE will take care of the rest.  What it does mean is that there is more boiler-plate code being written.  I'm also not sure what this does to versioning.  With the plan as it's written here, the interface doesn't need to change to support new events - only the GC and EE implementations need to change.

Calling all the way through to FireEtw* via a one-function-per-event interface would also mean that you would not need to communicate and store event handles because they would be considered an internal implementation detail of the EE, which is good.  This plan does, however, further cement the plan that adding new events that aren't custom events requires a rebuild and re-deployment of the EE and GC.

This is worth considering.

Reply:
Agreed - having read everyone's comments on this, I do also feel that it would be a substantial drop in complexity to have one callback per event, at the cost of increasing the number of callbacks on the interface.